design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,482 @@
+  - name: engineVersion
+    required: false
+    fieldPaths:
+    - ".spec.engineVersion"
+```
+
+The above WorkloadDefinition and Component leverages the existing Crossplane
+RedisCluster resource claim. Crossplane as a tentative OAM runtime need not
+define an extended workload type; it need only create a WorkloadDefinition
+pointing to [the existing CustomResourceDefinition] of the RedisCluster resource
+claim.
+
+The Component design enables OAM runtimes to break the task of reconciling the
+desired state of an application with the actual state of the world into two
+distinct stages:
+
+1. Render workloads. The OAM runtime iterates over the Components referenced by
+   an ApplicationConfiguration, instantiating a distinct workload resource for
+   each component. In the above examples this would involve creating a distinct
+   Server resource, and a distinct RedisCluster resource.
+2. Reconcile workloads. By producing valid instances of existing Kubernetes
+   custom resources OAM is able to leverage existing controllers. For example
+   the existing Crossplane RedisClaim controller(s) would be responsible for
+   reconciling the RedisCluster custom resource produced during the render
+   phase.
+
+Note that this two phase reconciliation and the production of distinct workload
+resources would not be required by the OAM specification. Runtimes could choose
+to iterate over the referenced components, render them in memory, and act upon
+the rendered components directly.
+
+### Traits
+
+Like workloads, each trait is of a distinct kind. A TraitDefinition defines the
+schema of each kind of trait, and limits what workload kinds it applies to:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: TraitDefinition
+metadata:
+  name: manualscalertrait.core.oam.dev
+spec:
+  appliesToWorkloads:
+    - server.core.oam.dev/v1alpha1
+    - worker.core.oam.dev/v1alpha1
+    - task.core.oam.dev/v1alpha1
+  definitionRef:
+    name: manualscalertrait.core.oam.dev
+```
+
+The above TraitDefinition references a Kubernetes CustomResourceDefinition named
+manualscalertrait.core.oam.dev, designating custom resources defined by said CRD
+to be valid OAM traits. As with WorkloadDefinitions, OAM runtimes that are not
+powered by Kubernetes could opt to either implement the CRD concept, or
+introduce a new field to the TraitDefinition spec that defines a kind of trait
+and its schema. Application operators can discover available OAM trait kinds by
+running `kubectl get traitdefinitions`. Traits would most frequently be defined
+by OAM runtimes or infrastructure operators.
+
+This document proposes that trait configuration be applied by be embedding one
+or more of a particular kind of trait under the `traits` array of the component
+it should apply to. For example:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ApplicationConfiguration
+metadata:
+  name: cool-example
+spec:
+  components:
+  - componentName: example-server
```

Comment:
why we don't need `instanceName` here?

Reply:
as I was reading the google doc, instance name seems to have been put under parameterValues